rtos/Mutex.cpp

```diff
@@ -50,30 +51,51 @@ void Mutex::constructor(const char *name)
     MBED_ASSERT(_id);
 }
 
-osStatus Mutex::lock(uint32_t millisec) {
+void Mutex::lock(void)
```

Comment:
Need a final decision on this return type, as per previous comment. @deepikabhavnani - any views?

Reply:
I thought we agreed no breaking change, thus revert this one, and add documentation note about return

Reply:
I'm fine with that, but still have a slight preference for just voiding it, so looking for 3rd or 4th votes just to make sure...

Reply:
The proposal was to have this separate `lock()` overload with the assert, but have it returning `osStatus` to minimise the compilation breakage. Documentation should note that it asserts and that use of the return value is deprecated, and it's planned to become void in 6.0. But that would have to be just as a note - you couldn't do it as a "MBED_DEPRECATED".

Reply:
Thanks! Something like this https://github.com/ARMmbed/mbed-os/pull/7423/commits/523478792ac44a656b81501ebba9a38cb35f85be . 

I would like to have both versions ready since tomorrow I will be on the road to Croatia ;-)

Reply:
I would vote for return instead of Asserts, in case of timeouts user should have control of software if they would like to perform cleanup and do some other activity instead of asserting?
Asserts are good idea for debug build but for stable OS we need other ways to deal with errors. 
I would prefer my OS to not break if I was not able to perform a specific task in one thread.

Reply:
The problem is that `mutex::lock()` failure is too fundamental and unexpected to realistically recover from. If your mutexes can't lock you have a real system problem, and it's not clear how you could recover.

Returning status from it is somewhat analogous to `*p = 3;` returning an error if `p` is an invalid pointer, rather than faulting.

And the error return would have to cascade up - the simplest `get_state()` method would need an error return if expected to "recover" from internal mutex lock failure.

See also the JIRA and issues linked in the description - this PR is the result of long previous discussion. No in-tree user is taking any action on mutex.lock() failure now. Either they ignore it, or they assert or error themselves. They have no "plan b" for a failure return.

Reply:
Return status, will be useful for release profile. Users who are vigilant and do check error status will have advantage in case of release profiles for issues which are not caught in debug mode.

Consider a case where application has to spawn multiple threads one for each activity and due to overload (denial of service kind of activity) if mutex lock timeouts, they will not prefer whole system to crash or proceed with timed out lock. Recovery for that application will be to kill child threads. 

> No in-tree user is taking any action on mutex.lock() failure now.
Does our cloud client application check return state?

Reply:
> No in-tree user is taking any action on mutex.lock() failure now.

I see your point here, checked Mbed OS code, we do not check error status. I am still of view returning error status should not be removed even if we assert. Corner cases and bugs without error proper handling are difficult to debug in Release builds.

Reply:
Mutex locks cannot timeout via this API. This is the no-timeout `lock` API, for mutex claims that are expected to work. The separate `try_lock` API returns a boolean indicating whether it timed out or not.

An application of the type you hypothesise - one that expects to kill threads still holding mutexes if they timeout - would not be using this API to do it.

Part of the motivation here is to clearly distinguish `lock`, which always works, from `try_lock`, which may not.

> Does our cloud client application check return state?

Cloud client core doesn't use the RTOS, except via the in-tree ns_event_loop mechanisms.  The PAL stuff for the extension bits uses CMSIS-RTOS APIs, rather than the mbed OS wrapper, afaict.

> I am still of view returning error status should not be removed even if we assert. Corner cases and bugs without error proper handling are difficult to debug in Release builds.

Maybe it should be an `error` rather than an assert, so it still traps in release builds (as pointer failures do).

Reply:
Maybe it should be an error rather than an assert, so it still traps in release builds (as pointer failures do).

👍

Reply:
I would MBED_ASSERT should call error internally not die.